### DISCUSSION TITLE
dynamic CFG_HOLDER value

### DISCUSSION THREAD

**Discussion by prabbit237** [arendst/Tasmota#18443] [Ideas]
This is actually more of a generic idea/question and I'm not as versed in PIO .ini files and/or Makefiles as I'd like to be but.....

I could see where having a CFG_HOLDER that auto-increments could be useful for self-compiled/testing versions. If there is a way to add something to the .ini file like:

````
-DCFG_HOLDER=((DAYS_SINCE_EPOCH_TIME-18000)*24)+CURRENT_HOUR_UTC)
````
then every time the hour that you compile it changes, it automatically bumps the counter up by 1 (the -18000 keeps it from overflowing, although a "MOD 32000" would also allow it to keep changing and just roll over) and thus the binary will usually load new settings from the SECTION 1 config (if needed, use minutes insteead of hours.) Or if the define is removed from the .ini file, it would simply use whatever number is in the my_user_config.h or the user_config_override.h the way it does now.

Can something be done in the .ini file to have a variable #define like that? I know the .ini file can basically act as a simple script file to decide what to compile, where to find the files, etc but can it perform functions like math, etc to create #defines similar to a bash shell script? From https://tech.aabouzaid.com/2022/09/create-makefile-targets-with-dynamic-parameters-and-autocompletion.html, it looks like a Makefile can be created that could do something like that using the $(shell...) line. But can it be done in the .ini as well?

And yes, this may be so special of a use-case that it's not worth adding it to the main code but if someone can say "yeah, check http://...... for an example of a dynamic value in a PIO .INI file", I could probably work it out myself. I'm just trying to find out if it's even possible before trying to search all over only to find I'm wasting my time.

**Comment by Jason2866** [comment]
It is a bad idea to change `CFG_HOLDER` variable. By doing this every setting will be deleted when doing an upgrade. This will NOT be implemented.

You can enhance Platformio as described here https://docs.platformio.org/en/latest/scripting/index.html
See folder `pio-tools` in Tasmota for examples

**Reply by prabbit237** [reply]
No, I wasn't expecting it to be implemented as a any sort of a default thing, definitely, but wanted it more as a debugging/very-special-case option that could be plugged into an over-ride. An "erase and flash" or "OTA and reset" would have the same effect but I was going for something a bit more automatic when desired. Then if I got other things setup in a way that I'm satisfied with it, I can then comment out a line or two and have it quit changing the variable for production runs.

https://docs.platformio.org/en/latest/scripting/examples/asking_for_input.html looks like it's the sort of principle that'd work to set a dynamic define and that's the main thing I was going for was just where to start the search. Thanks.

**Comment by Jason2866** [comment]
You can change everything like you want. If you provide a PR AND there is a brider use case the PR will be checked and considered to be merged.
